sdk/src/Core/Amazon.Runtime/Credentials/FallbackCredentialsFactory.cs

```diff
@@ -124,16 +131,22 @@ public static AWSCredentials GetCredentials()
 
         public static AWSCredentials GetCredentials(bool fallbackToAnonymous)
         {
-            if (cachedCredentials != null)
-                return cachedCredentials;
-
+            cachedCredentialsLock.EnterReadLock();
+            var credentials = cachedCredentials;
+            cachedCredentialsLock.ExitReadLock();
+            
+            if (credentials != null)
+            {
+                return credentials;
+            }
+            
             List<Exception> errors = new List<Exception>();
 
             foreach (CredentialsGenerator generator in CredentialsGenerators)
```

Comment:
This whole loop should be avoided running in multiple threads. Only one thread should be hitting EC2 instance metadata or parsing through profile config files to determine credentials. 

I suggest:
* Move the `cachedCredentialsLock.EnterWriteLock()` above this for loop.
* Start a `try` block
* Once in the block check again to see if `cachedCredentials` has a value if so return. It could have a value if this is the second thread through the write block and the first thread obtain the credentials. Here is an another example in the SDK doing the lock double check https://github.com/aws/aws-sdk-net/blob/9a993b556bd6a9ad2e06ce9799bbda1b99a56b69/sdk/src/Core/RegionEndpoint/RegionEndpointProviderV3.cs#L470
* Run the `foreach`.
* After the foreach `cachedCredentials = credentials`
* Close the `try` block and add a `finally` block that calls `cachedCredentialsLock.ExitWriteLock()` to exit the write lock

I think actually if you do put the whole `foreach` in the write lock you can get rid of the temporary `credentials` variable and simplify the logic since access to the `cachedCredentials` variable would now be controlled through the reader writer lock.